File: javascript/postgresql.js
"use strict";

const { Client } = require("pg"); // npm i pg
require("dotenv").config({ path: "../.env" }); // npm i dotenv

class Connector {
  constructor() {
    (this.database = process.env.DATABASE),
      (this.host = process.env.HOST),
      (this.user = process.env.DB_USER),
      (this.password = process.env.DB_PASSWORD);
  }

  connectToDatabase = () => {
    /* Establish a connection to the database
       and return a client object */
    const client = new Client({
      host: this.host,
      port: 5432,
      database: this.database,
      user: this.username,
      password: this.password,
      idle_in_transaction_session_timeout: 3,
    });
    return client;
  };
}

class User extends Connector {
  getUser = async (argument) => {
    const client = this.connectToDatabase();
    await client.connect();
    try {
      const query = await client.query(`SELECT * FROM users
                                        WHERE username = '${argument}';`);
      await client.end();
      return query.rows;
    } catch (err) {
      console.log(err);
    }
  };

  insertUser = async (username, age) => {
    const client = this.connectToDatabase();
    await client.connect();
    try {
      await client.query(`INSERT INTO users (username, age)
                          VALUES ('${username}', '${age}');`);
      await client.end();
    } catch (err) {
      console.log(err);
      await client.end();
    }
  };
}

const test = new User();
test.insertUser("Mavis", 3);
test.getUser("Stoffel").then((res) => console.log(res));

File: python/databases/postgresql.py
import psycopg2                     # pip install psycopg2
from dotenv import load_dotenv      # pip install python-dotenv
import os

load_dotenv(dotenv_path='../../.env')


class Connector:

    def __init__(self) -> None:

        self.user = os.getenv("DB_USER")
        self.password = os.getenv("DB_PASSWORD")
        self.database = os.getenv("DATABASE")
        self.host = os.getenv("HOST")

    def connect_to_database(self):
        """
        Establishes and active connection to the database and returns to connection
        """
        conn = psycopg2.connect(
            database=self.database,
            host=self.host,
            user=self.user,
            password=self.password
        )

        if conn:
            return conn
        else:
            raise Exception("Database connection failed")


class Users(Connector):

    def get_all_users(self):
        conn = self.connect_to_database()
        curr = conn.cursor()

        try:
            curr.execute('SELECT * FROM users')
            query = curr.fetchall()
            conn.commit()
            conn.close()
            return (query)  # return a list of tuples
        except Exception as e:
            print(e)
            conn.rollback()
            conn.close()

    def get_user(self, username):
        conn = self.connect_to_database()
        curr = conn.cursor()

        try:
            curr.execute(f"""SELECT * FROM users
                         WHERE username = '{username}'""")
            query = curr.fetchone()
            conn.commit()
            conn.close()
            return (query)  # return a tuple
        except Exception as e:
            print(e)
            conn.rollback()
            conn.close()

    def insert_user_record(self, data: tuple):
        conn = self.connect_to_database()
        cur = conn.cursor()
        try:
            cur.execute("""
                INSERT INTO users (username, age)
                VALUES (%s, %s);
                        """, data)
        except Exception as e:
            print(e)
        finally:
            conn.commit()
            conn.close()


if __name__ == "__main__":
    db = Users()
    db.connect_to_database()
    db.insert_user_record(("Rits", 2))
    print(db.get_all_users())
    print(db.get_user('Stoffel'))
